feat(android): getter/setter for applicationLocales in Android 12 and up

AbdullahFaqeir opened the pull request:
feat(android): via app compat delegate set application locales.
feat(android): via app compat delegate get application locales.
feat(android): via app compat delegate get default night mode
feat(android): via app compat delegate check if compat vector from resources enabled
feat(android): via app compat delegate set compat vector from recourses enabled or not.

**Description:**

This PR was raised from an issue appeared on android level 33 and later which caused Ti.Locale.setLanguage() not to work properly, which caused random pages not to be translated unless you recall Ti.Locale.setLanguage() on each page.

See [here](https://developer.android.com/guide/topics/resources/app-languages#java)

Nonetheless, I've ported the rest of static methods from AppCompatDelegate.

_edited by Michael:_ for a test code check https://github.com/tidev/titanium-sdk/pull/14008#issuecomment-2009179503

AbdullahFaqeir committed:
feat(android): via app compat delegate set default night mode
feat(android): via app compat delegate get default night mode
feat(android): via app compat delegate check if compat vector from resources enabled
feat(android): via app compat delegate set compat vector from recourses enabled or not.
feat(android): via app compat delegate set application locales.
feat(android): via app compat delegate get application locales.

m1ga commented:
Thanks for the PR! A few remarks:

* please import `androidx.appcompat.app.AppCompatDelegate;` so you don't need to fully put the path in there all the time as you can see in other files https://github.com/search?q=repo%3Atidev%2Ftitanium-sdk%20AppCompatDelegate&type=code

* use a different file to put your code in. E.g. the normal `setLanguage` is here: https://github.com/tidev/titanium-sdk/blob/e4cb309372361c325f48fde657291ad8d7eb1821/android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java#L215-L233 so you could duplicate that and adjust it for `setAppLanguage` . That way you can use `Ti.Locale.appLanguage = ""` later on. Make it just a setProperty/getProperty. 

* keep it for the locales parts only. E.g. `isCompatVectorFromResourcesEnabled` shouldn't be needed at all as it checks for <21 and the minSDK of Ti is 21. If it's only the locale stuff it makes it easier to test :wink: 

* `getDefaultNightMode` is in `Ti.UI.overrideUserInterfaceStyle` already.

* add some test code for the reviewer
* update `apidoc/Titanium/Locale/Locale.yml` and add the properties

AbdullahFaqeir committed:
feat(android): ti locale new property applicationLocales

m1ga commented on android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java:
```diff
@@ -51,6 +56,58 @@ public String getCurrentLocale()
 		return TiPlatformHelper.getInstance().getLocale();
 	}
 
+	@TargetApi(33)
+	@Kroll.method
+	@Kroll.setProperty
+	public void setApplicationLocales(String locales)
+	{
+		LocaleListCompat appLocale = LocaleListCompat.forLanguageTags(locales);
+		androidx.appcompat.app.AppCompatDelegate.setApplicationLocales(appLocale);
+	}
```
```suggestion
	@Kroll.setProperty
	public void setApplicationLocales(String locales)
	{
		if (Build.VERSION.SDK_INT >= 33) {
			LocaleListCompat appLocale = LocaleListCompat.forLanguageTags(locales);
			AppCompatDelegate.setApplicationLocales(appLocale);
		} else {
			Log.w(TAG, "Only available for Android 33 and later");
		}
	}
```

Only use `@Kroll.setProperty` and add the SDK check in the method, so it will show a warning on older devices but is still available

m1ga commented on android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java:
```diff
@@ -51,6 +56,58 @@ public String getCurrentLocale()
+	@Kroll.getProperty
+	public KrollDict[] getApplicationLocales()
+	{
+		LocaleListCompat localeListCompat = androidx.appcompat.app.AppCompatDelegate.getApplicationLocales();
+		int size = localeListCompat.size();
+		KrollDict[] locales = new KrollDict[size];
+		for (int i = 0; i < size; i++) {
+			Locale locale = localeListCompat.get(i);
+			if (locale != null) {
+				KrollDict localeObj = new KrollDict();
+				localeObj.put("country", locale.getCountry());
+				localeObj.put("iso3_country", locale.getISO3Country());
+				localeObj.put("display_country", locale.getDisplayCountry());
+				localeObj.put("language", locale.getLanguage());
+				localeObj.put("iso3_language", locale.getISO3Language());
+				localeObj.put("display_language", locale.getDisplayLanguage());
+				localeObj.put("variant", locale.getVariant());
+				localeObj.put("display_variant", locale.getDisplayVariant());
+				localeObj.put("script", locale.getScript());
+				localeObj.put("display_script", locale.getDisplayScript());
+				localeObj.put("display_name", locale.getDisplayName());
+				localeObj.put("language_tag", locale.toLanguageTag());
+				Character[] extensionKeys = new Character[locale.getExtensionKeys().size()];
+				String[] extensions = new String[locale.getExtensionKeys().size()];
+				Iterator<Character> extensionKeysSize = locale.getExtensionKeys().iterator();
+				int l = 0;
+				while (extensionKeysSize.hasNext()) {
+					extensionKeys[l] = extensionKeysSize.next();
+					extensions[l] = locale.getExtension(extensionKeys[l]);
+					l++;
+				}
+				localeObj.put("extension_keys", extensionKeys);
+				localeObj.put("extensions", extensions);
+				locales[i] = localeObj;
+			} else {
+				locales[i] = null;
+			}
+		}
+		return locales;
+	}
```
```suggestion
	@Kroll.getProperty
	public KrollDict[] getApplicationLocales()
	{
		if (Build.VERSION.SDK_INT < 33) {
			Log.w(TAG, "Only available for Android 33 and later");
			return null;
		}

		LocaleListCompat localeListCompat = AppCompatDelegate.getApplicationLocales();
		int size = localeListCompat.size();
		KrollDict[] locales = new KrollDict[size];
		for (int i = 0; i < size; i++) {
			Locale locale = localeListCompat.get(i);
			if (locale != null) {
				KrollDict localeObj = new KrollDict();
				localeObj.put("country", locale.getCountry());
				localeObj.put("iso3_country", locale.getISO3Country());
				localeObj.put("display_country", locale.getDisplayCountry());
				localeObj.put("language", locale.getLanguage());
				localeObj.put("iso3_language", locale.getISO3Language());
				localeObj.put("display_language", locale.getDisplayLanguage());
				localeObj.put("variant", locale.getVariant());
				localeObj.put("display_variant", locale.getDisplayVariant());
				localeObj.put("script", locale.getScript());
				localeObj.put("display_script", locale.getDisplayScript());
				localeObj.put("display_name", locale.getDisplayName());
				localeObj.put("language_tag", locale.toLanguageTag());
				Character[] extensionKeys = new Character[locale.getExtensionKeys().size()];
				String[] extensions = new String[locale.getExtensionKeys().size()];
				Iterator<Character> extensionKeysSize = locale.getExtensionKeys().iterator();
				int l = 0;
				while (extensionKeysSize.hasNext()) {
					extensionKeys[l] = extensionKeysSize.next();
					extensions[l] = locale.getExtension(extensionKeys[l]);
					l++;
				}
				localeObj.put("extension_keys", extensionKeys);
				localeObj.put("extensions", extensions);
				locales[i] = localeObj;
			} else {
				locales[i] = null;
			}
		}
		return locales;
	}
```

same here. Just `@Kroll.getProperty` and the SDK check inside.

m1ga commented on android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java:
```diff
@@ -21,8 +23,11 @@
 import org.appcelerator.titanium.util.TiPlatformHelper;
 import org.appcelerator.titanium.util.TiRHelper;
 
+import android.annotation.TargetApi;
```
```suggestion
import android.os.Build;
```

m1ga commented on android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java:
```diff
@@ -21,8 +23,11 @@
+import android.annotation.TargetApi;
 import android.telephony.PhoneNumberUtils;
 
+import androidx.core.os.LocaleListCompat;
```
```suggestion
import androidx.appcompat.app.AppCompatDelegate;
import androidx.core.os.LocaleListCompat;
```

m1ga reviewed (commented):
Added some code suggestions you can add in github with some minor code changes to adjust it to the rest of the SDK. It looks like much because I've added and if around it :smile: 

```js
Ti.Locale.applicationLocales = "de";
console.log(Ti.Locale.applicationLocales);
```

as some demo code

m1ga committed:
Merge branch 'master' into master

m1ga commented on tests/Resources/ti.locale.test.js:
```diff
@@ -44,6 +44,12 @@ describe('Titanium.Locale', () => {
 		});
 
+		describe.android('.applicationLocales', () => {
+			it('is a Array', () => {				
```
```suggestion
			it('is a Array', () => {
```

removing some spaces at the end :)

m1ga commented:
@AbdullahFaqeir you have to click the `Commit suggestions` buttons to apply the code or apply it locally and do another push. The "Resolve conversation" button will only close it but not apply it :)

AbdullahFaqeir committed:
Update android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java

Co-authored-by: Michael Gangolf <m1ga@users.noreply.github.com>

AbdullahFaqeir committed:
Update android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java

Co-authored-by: Michael Gangolf <m1ga@users.noreply.github.com>

AbdullahFaqeir committed:
Update android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java

Co-authored-by: Michael Gangolf <m1ga@users.noreply.github.com>

AbdullahFaqeir committed:
Update android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java

Co-authored-by: Michael Gangolf <m1ga@users.noreply.github.com>

AbdullahFaqeir committed:
Update android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java

Co-authored-by: Michael Gangolf <m1ga@users.noreply.github.com>

AbdullahFaqeir committed:
Update android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java

Co-authored-by: Michael Gangolf <m1ga@users.noreply.github.com>

AbdullahFaqeir committed:
Update android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java

Co-authored-by: Michael Gangolf <m1ga@users.noreply.github.com>

AbdullahFaqeir committed:
Update android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java

Co-authored-by: Michael Gangolf <m1ga@users.noreply.github.com>

AbdullahFaqeir commented:
> @AbdullahFaqeir you have to click the `Commit suggestions` buttons to apply the code or apply it locally and do another push. The "Resolve conversation" button will only close it but not apply it :)

Done :)

AbdullahFaqeir committed:
Update tests/Resources/ti.locale.test.js

Co-authored-by: Michael Gangolf <m1ga@users.noreply.github.com>

m1ga commented on android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java:
```diff
@@ -21,8 +23,16 @@
+import android.os.Build;
 import android.telephony.PhoneNumberUtils;
 
+import androidx.appcompat.app.AppCompatDelegate;
+import androidx.appcompat.app.AppCompatDelegate;
+import androidx.appcompat.app.AppCompatDelegate;
+import androidx.appcompat.app.AppCompatDelegate;
+import androidx.appcompat.app.AppCompatDelegate;
```
```suggestion
import androidx.appcompat.app.AppCompatDelegate;
```

strange, it added it multiple times. I'll fix that

m1ga committed:
Update android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java

m1ga commented:
It works but it's a bit tricky to use as it will restart the activity. I had an endless loop after I've added two setters below each other.

Here is a working demo
```js
var currentLang = Ti.App.Properties.getString("lang", "de");

const win = Ti.UI.createWindow({});
const btn = Ti.UI.createButton({title: "switch lang"});
const lbl1 = Ti.UI.createLabel({top: 10,text: currentLang});
const lbl = Ti.UI.createLabel({top: 30,text: L("word")});
win.add([btn, lbl, lbl1]);

btn.addEventListener("click", function() {
  currentLang = (currentLang == "de") ? "en" : "de";
  Ti.App.Properties.setString("lang", currentLang);
	Ti.Locale.applicationLocales = currentLang;
})

win.addEventListener("open", function() {
  // is called when you switch the language
  //
  currentLang = Ti.App.Properties.getString("lang", "de");
  console.log(Ti.Locale.applicationLocales)
  lbl1.text = currentLang;
  lbl.text = L("word");
})

win.open();
```
app/i18n/de/strings.xml
```
<?xml version="1.0" encoding="UTF-8"?>
<resources>
	<string name="word">DE</string>
</resources>
```
app/i18n/en/strings.xml
```
<?xml version="1.0" encoding="UTF-8"?>
<resources>
	<string name="word">EN</string>
</resources>
```


one last step @AbdullahFaqeir :
Edit https://github.com/tidev/titanium-sdk/blob/master/apidoc/Titanium/Locale/Locale.yml and add your
```
- name: applicationLocales
...
```
before: https://github.com/tidev/titanium-sdk/blob/master/apidoc/Titanium/Locale/Locale.yml#L221-L227

Summary something like `Sets the UI locales for the app`
Description: `You can set the UI locales using a value such as "en" or "fr". See the [ISO 639-1](http://en.wikipedia.org/wiki/ISO_639-1) section of wikipedia for reference. The getter will return an array of objects. 
Note: This is only available on Android 33 and up. It will also restart your current activity when you switch your language, the open event is called again.`

and this:
```
platforms: [android]
since: "12.4.0"
```

m1ga commented on android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java:
```diff
@@ -51,6 +57,63 @@ public String getCurrentLocale()
 	}
 
+	@Kroll.setProperty
+	public void setApplicationLocales(String locales)
```
```suggestion
	public void setApplicationLocale(String locales)
```

m1ga committed:
Update android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java

AbdullahFaqeir committed:
Update Locale.yml for PR #14008

Added documentation for applicationLocale and applicationLocales which are only available on Android level 33 and up.

m1ga commented on apidoc/Titanium/Locale/Locale.yml:
```diff
@@ -217,7 +217,44 @@ properties:
+    summary: Set the UI locale for the app.
+    description: |
+        Set the UI current active locale for the app.
+        When changed, it automatically causes the activity to be restarted.
```
```suggestion
        When changed, it automatically causes the activity to be restarted.
        Note: This is only available on Android 33 and up.
```

m1ga committed:
Update apidoc/Titanium/Locale/Locale.yml

m1ga reviewed (approved):
LGTM :+1: 

Thank you for the work

AbdullahFaqeir committed:
fix(android): add .idea to .gitignore,
chore(android): let there be gradle 8.

m1ga commented:
@AbdullahFaqeir careful: you've pushed the gradle part into this PR. Please revert the last changes.
You could also close it and start a fresh one in a different branch, not the master branch. Then you can work on two PRs at the same time.

AbdullahFaqeir commented:
> @AbdullahFaqeir careful: you've pushed the gradle part into this PR. Please revert the last changes. You could also close it and start a fresh one in a different branch, not the master branch. Then you can work on two PRs at the same time.

Yeah I did that by mistake, I forgot to close this PR, I'll fix it rightaway.

AbdullahFaqeir commented:
Reopening.
